substrate/frame/assets/src/migration/v2.rs

```diff
@@ -0,0 +1,191 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{Account, Asset, AssetDetails, AssetStatus, Config, DepositBalanceOf, Pallet, Vec};
+use frame_support::{
+	pallet_prelude::*,
+	sp_runtime::traits::{Saturating, Zero},
+	traits::OnRuntimeUpgrade,
+	weights::Weight,
+};
+
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+
+pub mod old {
+	use super::*;
+
+	#[frame_support::storage_alias]
+	/// Details of an asset.
+	pub(crate) type Asset<T: Config<I>, I: 'static> = StorageMap<
+		Pallet<T, I>,
+		Blake2_128Concat,
+		<T as Config<I>>::AssetId,
+		AssetDetails<
+			<T as Config<I>>::Balance,
+			<T as frame_system::Config>::AccountId,
+			DepositBalanceOf<T, I>,
+		>,
+	>;
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub(crate) struct AssetDetails<Balance, AccountId, DepositBalance> {
+		pub owner: AccountId,
+		pub issuer: AccountId,
+		pub admin: AccountId,
+		pub freezer: AccountId,
+		pub supply: Balance,
+		pub deposit: DepositBalance,
+		pub min_balance: Balance,
+		pub is_sufficient: bool,
+		pub accounts: u32,
+		pub sufficients: u32,
+		pub approvals: u32,
+		pub status: AssetStatus,
+	}
+
+	impl<Balance, AccountId, DepositBalance> AssetDetails<Balance, AccountId, DepositBalance>
+	where
+		Balance: Zero,
+	{
+		pub(super) fn migrate_to_v2(
+			self,
+		) -> super::AssetDetails<Balance, AccountId, DepositBalance> {
+			super::AssetDetails {
+				owner: self.owner,
+				issuer: self.issuer,
+				admin: self.admin,
+				freezer: self.freezer,
+				supply: self.supply,
+				inactive: Zero::zero(),
+				deposit: self.deposit,
+				min_balance: self.min_balance,
+				is_sufficient: self.is_sufficient,
+				accounts: self.accounts,
+				sufficients: self.sufficients,
+				approvals: self.approvals,
+				status: self.status,
+			}
+		}
+	}
+}
+
+/// This migration moves all the state to v2 of Assets
+pub struct VersionUncheckedMigrateToV2<
+	T: Config<I>,
+	I: 'static,
+	A: Get<Vec<(T::AssetId, T::AccountId)>>,
+>(sp_std::marker::PhantomData<(T, I, A)>);
+
+impl<T: Config<I>, I: 'static, A: Get<Vec<(T::AssetId, T::AccountId)>>> OnRuntimeUpgrade
+	for VersionUncheckedMigrateToV2<T, I, A>
+{
+	fn on_runtime_upgrade() -> Weight {
+		let mut translated = 0u64;
+
+		Asset::<T, I>::translate::<
+			old::AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>,
+			_,
+		>(|_asset_id, old_value| {
+			translated.saturating_inc();
+			Some(old_value.migrate_to_v2())
+		});
+
+		let mut reads = 0u64;
+		for (asset_id, account) in A::get() {
+			reads.saturating_inc();
+			let Some(acc) = Account::<T, I>::get(&asset_id, &account) else {
+				log::info!(
+					"inactive migration: account {:?} not found for asset {:?}",
+					account,
+					asset_id
+				);
+				continue
+			};
+			Asset::<T, I>::mutate(&asset_id, |asset| match asset {
+				Some(asset) => asset.inactive.saturating_accrue(acc.balance),
```

Comment:
Is it correct to set everything as inactive? If so, couldn't we use the total asset issuance instead of looping through all accounts?

Reply:
The parameter `A: Vec<(AssetId, AccountId)>` represents the inactive balances for each asset. 
The method doesn't loop through all accounts, but only through the inactive accounts for certain assets which are passed as inputs.